Paginate the results from `boto.iam`

jsternberg opened the pull request:
The default pagination is every 100 items with a maximum of 1000 from
Amazon. This properly uses the marker returned by Amazon to concatenate
the various pages from the results.

This fixes #2440.

bcoca commented:
i like the generic paginate function, i might copy it to the shared ec2.py and allow for most other modules to use it

bcoca commented:
fixes http://github.com/ansible/ansible/issues/13242

Etherdaemon commented:
Should the user get a choice in the number of items to get returned though, rather than always returning all results? Also in boto3 the marker syntax is slightly different between different AWS types. EG EC2 has NextToken while Cloudfront has Marker.

bcoca commented:
the option would have to be added at the module level (probably a common spec like most modules already have, defaulting to all)

stmarier commented:
in my opinion there's no reason to have a choice on the number of items returned, at least (or especially) with the iam module.

I have a similar fix for this already running in our fork of this repo because currently what happens is if you have over 100 iam users, the module incorrectly assumes that users after 100 don't exist and tries to create them, then you get an unhelpful error about <user> already exists and the module fails

so I'm not really sure what you would accomplish by adding a max_items parameter to the module except giving people the opportunity to break their stuff

bcoca commented:
@stmarier good point, the limit would only make sense in gathering IAM 'facts'

gregdek commented:
Thanks @jsternberg for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

gregdek commented:
Thanks @jsternberg for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

jsternberg commented:
Sure. Let me try rebasing it.

jsternberg committed:
Paginate the results from `boto.iam`

The default pagination is every 100 items with a maximum of 1000 from
Amazon. This properly uses the marker returned by Amazon to concatenate
the various pages from the results.

This fixes #2440.

ryansb commented:
Related to #2440

gregdek commented:
@jsternberg A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

jsternberg commented:
I've rebased it again.

ready_for_review

gregdek commented:
Thanks @jsternberg for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

ryansb commented:
:ship: looks good to me, shipit
